fdbclient/S3Cp.actor.cpp

```diff
@@ -0,0 +1,579 @@
+	if (pFiles != nullptr) {
+		for (auto& f : blobCredentials) {
+			pFiles->push_back(f);
+		}
+	}
+}
+
+bool S3CpTLSConfig::setupTLS() {
+	if (tlsCertPath.size()) {
+		try {
+			setNetworkOption(FDBNetworkOptions::TLS_CERT_PATH, tlsCertPath);
+		} catch (Error& e) {
+			std::cerr << "ERROR: cannot set TLS certificate path to " << tlsCertPath << " (" << e.what() << ")\n";
+			return false;
+		}
+	}
+
+	if (tlsCAPath.size()) {
+		try {
+			setNetworkOption(FDBNetworkOptions::TLS_CA_PATH, tlsCAPath);
+		} catch (Error& e) {
+			std::cerr << "ERROR: cannot set TLS CA path to " << tlsCAPath << " (" << e.what() << ")\n";
+			return false;
+		}
+	}
+	if (tlsKeyPath.size()) {
+		try {
+			if (tlsPassword.size())
+				setNetworkOption(FDBNetworkOptions::TLS_PASSWORD, tlsPassword);
+
+			setNetworkOption(FDBNetworkOptions::TLS_KEY_PATH, tlsKeyPath);
+		} catch (Error& e) {
+			std::cerr << "ERROR: cannot set TLS key path to " << tlsKeyPath << " (" << e.what() << ")\n";
+			return false;
+		}
+	}
+	if (tlsVerifyPeers.size()) {
+		try {
+			setNetworkOption(FDBNetworkOptions::TLS_VERIFY_PEERS, tlsVerifyPeers);
+		} catch (Error& e) {
+			std::cerr << "ERROR: cannot set TLS peer verification to " << tlsVerifyPeers << " (" << e.what() << ")\n";
+			return false;
+		}
+	}
+	return true;
+}
+
+// Get the endpoint for the given s3url.
+// Populates parameters and resource with parse of s3url.
+static Reference<S3BlobStoreEndpoint> getEndpoint(std::string s3url,
+                                                  std::string& resource,
+                                                  S3BlobStoreEndpoint::ParametersT& parameters) {
+	std::string error;
+	Reference<S3BlobStoreEndpoint> endpoint =
+	    S3BlobStoreEndpoint::fromString(s3url, {}, &resource, &error, &parameters);
+	if (resource.empty()) {
+		TraceEvent(SevError, "EmptyResource").detail("s3url", s3url);
+		throw backup_invalid_url();
+	}
+	for (auto c : resource) {
+		if (!isalnum(c) && c != '_' && c != '-' && c != '.' && c != '/') {
+			TraceEvent(SevError, "IllegalCharacter").detail("s3url", s3url);
+			throw backup_invalid_url();
+		}
+	}
+	if (error.size()) {
+		TraceEvent(SevError, "GetEndpointError").detail("s3url", s3url).detail("error", error);
+		throw backup_invalid_url();
+	}
+	return endpoint;
+}
+
+// Copy filepath to bucket at resource in s3.
+ACTOR Future<Void> copy_up_file(Reference<S3BlobStoreEndpoint> endpoint,
+                                std::string bucket,
+                                std::string resource,
+                                std::string filepath) {
+	// Reading an SST file fully into memory is pretty obnoxious. They are about 16MB on
+	// average. Streaming would require changing this s3blobstore interface.
+	// Make 32MB the max size for now even though its arbitrary and way to big.
+	state std::string content = readFileBytes(filepath, 1024 * 1024 * 32);
```

Comment:
what happens if file is larger than 32MB?

Reply:
readFileBytes will throw file_too_large exception
```

	size_t size = fileSize(filename);
	if (size > maxSize) {
		TraceEvent("ReadFileBytes_FileTooLarge")
		    .detail("FileSize", size)
		    .detail("InputMaxSize", maxSize)
		    .detail("Filename", filename);
		throw file_too_large();
	}

```

Reply:
(Thanks for reviewing) Would you like a different behavior?

Reply:
That should be good for testing. I just want to make sure we are not silently masking errors.